text/0000-assign-param-syntax.md

```diff
@@ -0,0 +1,113 @@
+
+Consider the following real-world code example from the `ponycc` repository, in which a simple data class with six fields requires 22 lines of code with redundant type information, just to get the class off the ground with a constructor and before writing any useful methods:
+
+```pony
+class val UseFFIDecl is (AST & UseDecl)
+  let attachments: (Attachments | None)
+  let name: (Id | LitString)
+  let return_type: TypeArgs
+  let params: Params
+  let partial: (Question | None)
+  let guard: (IfDefCond | None)
+
+  new val create(
+    name': (Id | LitString),
+    return_type': TypeArgs,
+    params': Params,
+    partial': (Question | None),
+    guard': (IfDefCond | None) = None,
+    attachments': (Attachments | None) = None)
+  =>
+    this.attachments = attachments'
+    this.name = name'
+    this.return_type = return_type'
+    this.params = params'
+    this.partial = partial'
+    this.guard = guard'
+```
+
+# Detailed design
+
+The change proposed in this RFC introduces "assign parameters", which allow a field to be referenced in the position of a parameter, indicating that the argument passed for that parameter position should be immediately assigned to the field rather than given its own name in the local scope.
+
+Each field will be referenced with its `this.`-prefixed identifier, just as it would be referenced in the body of a method (though unlike in that context, the `this.` prefix would be required rather than optional here). Here is the same example from above, rewritten now in just 9 lines compared to the earlier 22:
+
+```pony
+class val UseFFIDecl is (AST & UseDecl)
+  let attachments: (Attachments | None)
+  let name: (Id | LitString)
+  let return_type: TypeArgs
+  let params: Params
+  let partial: (Question | None)
+  let guard: (IfDefCond | None)
+
+  new val create(this.name, this.return_type, this.params, this.partial, this.guard, this.attachments)
+```
+
+Note that there is now no syntactical body of the method, but the compiler will add a body that does the appropriate assignments.
+
+Of course, "assign parameters" can be used alongside normal parameters, and not every field need be associated to an "assign parameter". Assign parameters can also have a "default value" as normal parameters can, which specifies the value to use if the caller does not supply an argument for that position. The following refactored example demonstrates all of those concepts:
+
+```pony
+class val UseFFIDecl is (AST & UseDecl)
+  let attachments: (Attachments | None)
+  let name: (Id | LitString)
+  let return_type: TypeArgs
+  let params: Params
+  let partial: (Question | None)
+  let guard: (IfDefCond | None)
+
+  new val create(name: String, this.return_type, this.params, this.attachments = None) =>
+    this.name = Id(name)
+    this.partial = None
+    this.guard = None
```

Comment:
What happens with this PR if the body of the constructor is now empty? Do we still need to include `=> None`?

Reply:
My expectation would be yes, you need to. I think it is worth explicitly noting that in this RFC.

Reply:
I've updated it to explicitly mention that requirement.